fv3core/utils/grid.py

```diff
@@ -366,14 +367,203 @@ def horizontal_starts_from_shape(self, shape):
+        return cls(
+            origin=grid.compute_origin(),
+            domain=grid.domain_shape_compute(),
+            n_halo=grid.halo,
+            south_edge=grid.south_edge,
+            north_edge=grid.north_edge,
+            west_edge=grid.west_edge,
+            east_edge=grid.east_edge,
+        )
+
+    @property
+    def isc(self):
+        """start of the compute domain along the x-axis"""
+        return self.origin[0]
+
+    @property
+    def iec(self):
+        """end of the compute domain along the x-axis"""
+        return self.origin[0] + self.domain[0]
+
+    @property
+    def jsc(self):
+        """start of the compute domain along the y-axis"""
+        return self.origin[1]
+
+    @property
+    def jec(self):
+        """end of the compute domain along the y-axis"""
+        return self.origin[1] + self.domain[1]
+
+    @property
+    def ks(self):
+        """start of the data domain along the z-axis"""
+        return self.origin[2]
+
+    @property
+    def ke(self):
+        """end of the data domain along the z-axis"""
+        return self.origin[2] + self.domain[2]
+
+    @property
+    def isd(self):
+        """start of the full domain including halos along the x-axis"""
+        return self.origin[0] - self.n_halo
+
+    @property
+    def ied(self):
+        """end of the full domain including halos along the x-axis"""
+        return self.isd + self.domain[0] + 2 * self.n_halo
+
+    @property
+    def jsd(self):
+        """start of the full domain including halos along the y-axis"""
+        return self.origin[1] - self.n_halo
+
+    @property
+    def jed(self):
+        """end of the full domain including halos along the y-axis"""
+        return self.jsd + self.domain[1] + 2 * self.n_halo
+
+    def origin_full(self, add: Index3D = (0, 0, 0)):
+        """
+        Returns the origin of the full domain including halos, plus an optional offset.
+        """
+        return (self.isd + add[0], self.jsd + add[1], self.ks + add[2])
+
+    def origin_compute(self, add: Index3D = (0, 0, 0)):
+        """
+        Returns the origin of the compute domain, plus an optional offset
+        """
+        return (self.isc + add[0], self.jsc + add[1], self.ks + add[2])
+
+    def domain_full(self, add: Index3D = (0, 0, 0)):
+        """
+        Returns the shape of the full domain including halos, plus an optional offset.
+        """
+        return (
+            self.ied - self.isd + add[0],
+            self.jed - self.jsd + add[1],
+            self.ke - self.ks + add[2],
+        )
+
+    def domain_compute(self, add: Index3D = (0, 0, 0)):
+        """
+        Returns the shape of the compute domain, plus an optional offset.
+        """
+        return (
+            self.iec - self.isc + add[0],
+            self.jec - self.jsc + add[1],
+            self.ke - self.ks + add[2],
+        )
+
 
-@functools.lru_cache(maxsize=None)
 def axis_offsets(
+    grid: Union[Grid, GridIndexing],
```

Comment:
should axis_offsets live in grid.py? since they take in a grid, maybe it should belong to the GridIndexing class. The logic has already been duplicated to work with the new class anyways.

Reply:
I like it. I'll move the new version to be a method on that class.